### ISSUE TITLE
State Exposed to Effects?

### ISSUE THREAD

**Issue by zebulonj** [FormidableLabs/freactal#29]
Is there a way to read state in effects, prior to the atomic state transformation?

In `redux`, a [prototypical example](http://redux.js.org/docs/advanced/AsyncActions.html#async-action-creators) would be:

```js
export function fetchPostsIfNeeded(subreddit) {
  // Note that the function also receives getState()
  // which lets you choose what to dispatch next.

  // This is useful for avoiding a network request if
  // a cached value is already available.

  return (dispatch, getState) => {
    if (shouldFetchPosts(getState(), subreddit)) {
      // Dispatch a thunk from thunk!
      return dispatch(fetchPosts(subreddit))
    } else {
      // Let the calling code know there's nothing to wait for.
      return Promise.resolve()
    }
  }
}
```

In that example, `getState()` is used to inspect whether a fetch is needed.

**Comment by VinSpee** [comment]
```es6
const wrapComponentWithState = provideState({
  initialState: () => ({
    givenName: "Walter",
    familyName: "Harriman",
    locale: "en-us"
  }),
  effects: {
    setGivenName: softUpdate((state, val) => ({ givenName: val })),
    setFamilyName: softUpdate((state, val) => ({ familyName: val }))
  },
  computed: {
    fullName: ({ givenName, familyName, locale }) => startsWith(locale, "en") ?
      `${givenName} ${familyName}` :
      `${familyName} ${givenName}`,
    greeting: ({ fullName, locale }) => startsWith(locale, "en") ?
      `Hi, ${fullName}, and welcome!` :
      `Helló ${fullName}, és szívesen!`
  }
});
```

**Comment by VinSpee** [comment]
I think something like this would work:

```es6
const wrapComponentWithState = provideState({
  effects: {
    shouldFetchPosts: (effects, subreddit, state) => { // ...your logic },
    fetchPosts: (effects, subreddit) => { // your logic },
    fetchPostsIfNeeded: (effects, subreddit) => state => {
      if (effects.shouldFetchPosts(state, subreddit)) { // you have the option of passing state here, or you could access it inside of your `fetchPosts` effect, depending on your needs
        effects.fetchPosts(subreddit);
      }
      return Promise.resolve();
   },
});
```

notice that effects return a function that is passed `state` as a parameter. See the [effect arguments](https://github.com/FormidableLabs/freactal#effect-arguments) for details.

**Comment by zebulonj** [comment]
Yeah. It's clear that arguments can be passed to the effects, and that state—exposed to the calling component via `injectState()`—could be one of those arguments. I'd prefer to hide some of that from the consumer of the state.

What I'm describing would probably have to be accomplished via middleware, same as `Redux` requires `redux-thunk` middleware to expose `getState()` to the action creator.

**Comment by zebulonj** [comment]
@VinSpee I just looked more closely at your proposed approach, and see now that it doesn't require the consumer to pass the state. Sorry for overlooking that.

A concern...

In your proposed `fetchPostsIfNeeded`, the following portion is the reducer:

```js
const reducer = state => {
  if (effects.shouldFetchPosts(state, subreddit)) { // you have the option of passing state here, or you could access it inside of your `fetchPosts` effect, depending on your needs
    effects.fetchPosts(subreddit);
  }
  return Promise.resolve();
}
```

However, `freactal` expects the reducer to a synchronous function that returns a transformed state. 
Effects have to take the form:
```js
const effect = ( effects, ...args ) => Promise.resolve( state => transform( state ) );
```
The `Promise.resolve( ... )` wrapping the synchronous state transformation can be implied, because of the way that effects are [wrapped](https://github.com/FormidableLabs/freactal/blob/master/src/effects.js#L9-L11) when the provider is created, but ultimately the reducer where state is received and transformed must be synchronous.

The `Promise` you are returning in your proposed effect would become the argument passed to `hocState.setState( ... )` via [`getEffects`](https://github.com/FormidableLabs/freactal/blob/master/src/effects.js) and the enclosed `applyReducer()`.

**Comment by divmain** [comment]
Confession: I originally passed both `effects` and `state` to each effect upon invocation, with the same thoughts you've expressed here.

However, I ultimately removed it since the promise-that-resolves-to-a-function thing is unfamiliar to most folks, and providing state both at the top of the effect and as an argument to the resolved-to-function could get confusing quite fast.

This might be an appropriate place for a helper function of some sort - something that makes it easier to build conditional effects.  Let me think on this a bit.  And if you have any thoughts on the matter, I'm all ears.  For now, there's a sufficient work-around in passing in state (or a piece of state) into the effect directly.

**Comment by markacola** [comment]
I have been using this helper.
```js
const pure = (fn) => (effects, ...args) => state => (fn(effects, state, ...args), state)
```
I don't know that the naming is the greatest, but happy to submit a quick pull request to add a helper.

**Comment by divmain** [comment]
@markacola, a PR would be great.  I think we'll want a different name though, since "pure" is quite the overloaded term at this point.

I assume you're using this `pure` function to wrap each effect?  The problem here is that the effect might return a promise, and the state-update function needs to be synchronous.

Another possibility would be to add this functionality as a middleware.  It would wrap each effect and add an additional `state` arg before the rest, much like you've done with the helper above.  I wouldn't have a problem with shipping this as a standard baked-in middleware.

**Comment by markacola** [comment]
Yes, I wrap each effect with this `pure` function. This does return a synchronous state-update function, and ignores whatever is returned from the `fn`, returning the original state instead.
The idea is that the effect can only affect the state by invoking other effects, so it is pure in that sense, but I cannot think of another name.
A middleware could be very clean. Would you want the middleware exposed from `helpers.js`?

**Comment by KwanMan** [comment]
@divmain @markacola I've been using this middleware

```js
const withStateMiddleWare = freactalContext => {
  const { state, effects } = freactalContext
  const effectsWithState = Object.keys(effects).reduce((memo, key) => {
    const baseEffect = effects[key]
    return {
      ...memo,
      [key]: (...args) => baseEffect(state, ...args)
    }
  }, {})
  return {
    ...freactalContext,
    effects: effectsWithState
  }
}
```

which changes the signature of an effect from:

```js
(effects, arg1, arg2) => {/* ... */}
```

to

```js
(effects, state, arg1, arg2) => {/* ... */}
```

giving you access to the state before deciding how to perform your action

I can PR this if you think it's a good solution

**Comment by cheapsteak** [comment]
Another use case for wanting access to state along with effects - We need to read the state to make an ajax request, then modify the state with values from the response

**Comment by divmain** [comment]
Here's what I'm thinking for the middleware.  For the time being, this should serve as a work-around for anyone here still wondering how to do this:

```javascript
import React, { Component } from "react";
import { render } from "react-dom";
import { provideState, injectState } from "freactal";


const supplyStateToEffects = freactalCxt => Object.assign({}, freactalCxt, {
  effects: Object.keys(freactalCxt.effects).reduce((memo, key) => {
    memo[key] = (...args) => freactalCxt.effects[key](freactalCxt.state, ...args);
    return memo;
  }, {})
});

const wrapComponentWithState = provideState({
  initialState: () => ({ counter: 0 }),
  middleware: [supplyStateToEffects],
  effects: {
    addOneUpToTen: (effects, injectedState, text) => {
      if (injectedState.counter > 9) { return null; }
      return state => Object.assign({}, state, { counter: state.counter + 1 })
    }
  }
});

const Parent = wrapComponentWithState(injectState(({ state, effects }) => (
  <div>
    {`Our counter is at: ${state.counter}`}
    <button onClick={effects.addOneUpToTen}>Add One</button>
  </div>
)));

render(<Parent />, document.getElementById("root"));
```

Keep in mind that the state passed into the effect is the computed state; it'll include state from all parent state containers.

@zebulonj @VinSpee @cheapsteak - does this address your need, and would it be helpful to include middleware like this along with the rest of `freactal`?

**Comment by ksvitkovsky** [comment]
To anyone trying to solve this challenge, note that when you use the middleware given above, state is not being passed to effects that are invoked via `effects`-argument. So you should either pass it explicitly or not expect it to be passed in those "inner" effects.

IMO, it would be better to enforce passing state to invoked effects in your project for consistency

```js
effects: {
  publicEffect: (effects, state, ...args) {
    effects.privateEffect(state, arg);
  },
  privateEffect: (effects, state, ...args) {}
}
```

@divmain I think the issue we discuss here is quite common and should at least be covered in README. Middleware could be handy as well (note the behaviour I described above) and changing API for effects to expect second argument to be state is even better

**Comment by bingomanatee** [comment]
This is doable, in two stages; you turn the state modifier into an identity function that also calls a second effect with parameters from state. Check out side effects in freactal-seed.
